File: app/src/main/java/com/jeksonshar/jweather/ui/WeatherFragment.java
package com.jeksonshar.jweather.ui;

import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.jeksonshar.jweather.R;
import com.jeksonshar.jweather.model.WeatherListModel;
import com.jeksonshar.jweather.model.WeatherModel;
import com.jeksonshar.jweather.repository.Repository;
import com.jeksonshar.jweather.repository.RepositoryProvider;
import com.jeksonshar.jweather.request.Networking;

import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class WeatherFragment extends Fragment {

    int city = 922137; // Kharkiv

    private RecyclerView mRecyclerView;
    private WeatherViewModel mWeatherViewModel;
    private WeatherAdapter mWeatherAdapter;
    private TextView lastUpdate;
    private ImageButton changeView;

    private List<WeatherModel> mWeatherModels;
    private Repository mRepository;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        mRepository = RepositoryProvider.getInstance(getContext());

        mWeatherViewModel = new ViewModelProvider(this, new ViewModelFactory(city))
                .get(WeatherViewModel.class);

//        new InternetRequestTask().execute();      // do it if not using ViewModel
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_weather, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        lastUpdate = view.findViewById(R.id.last_update_view);
        mRecyclerView = view.findViewById(R.id.recyclerView);
        changeView = view.findViewById(R.id.fab);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        SimpleDateFormat dateFormat = new SimpleDateFormat(
                " d MMM yyyy; HH:mm", Locale.getDefault());
        lastUpdate.setText(dateFormat.format(mWeatherViewModel.getLastUpdate().getTime()));

        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        mWeatherModels = RepositoryProvider.getInstance(getActivity()).getAllItemsWeather();
        if (mWeatherModels != null) {
            mWeatherAdapter = new WeatherAdapter(mWeatherModels);
            mRecyclerView.setAdapter(mWeatherAdapter);
        }

        mWeatherViewModel.getRequest().observe(getViewLifecycleOwner(), new Observer<List<WeatherModel>>() {
            @Override
            public void onChanged(List<WeatherModel> weatherModels) {
                mWeatherModels = weatherModels;
                if (!mWeatherModels.isEmpty()) {
                    mRepository.saveItemsInRoom(mWeatherModels);
                } else {
                    Toast.makeText(getContext(), R.string.no_internet, Toast.LENGTH_LONG)
                            .show();
                }
                mWeatherAdapter = new WeatherAdapter(RepositoryProvider.getInstance(getContext())
                        .getAllItemsWeather());
                mRecyclerView.setAdapter(mWeatherAdapter);
            }
        });

        changeView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //TODO
            }
        });
    }

//          // Do what is below if not using ViewModel - пока не удалять
//    @MainThread
//    private void applyItemWeathers() {
//        mWeatherAdapter = new WeatherAdapter(RepositoryProvider.getInstance(getContext()).getAllItemsWeather());
//        mRecyclerView.setAdapter(mWeatherAdapter);
//    }
//
//    @WorkerThread
//    private List<WeatherModel> executeRequest(){
//        WeatherListModel listWeatherModel = Networking.makeRequestByCity(city);
//        if (listWeatherModel != null) {
//            return listWeatherModel.getConsolidatedWeather();
//        }
//        else  {
//            return Collections.emptyList();
//        }
//    }
//
//    private class InternetRequestTask extends AsyncTask<Void, Void, List<WeatherModel>> {
//
//        @Override
//        protected List<WeatherModel> doInBackground(Void... voids) {
//            return executeRequest();
//        }
//
//        @Override
//        protected void onPostExecute(List<WeatherModel> weatherModels) {
//            if (!weatherModels.isEmpty()) {
//                mRepository.saveItemsInRoom(weatherModels);
//            } else {
//                try {
//                    Toast.makeText(getActivity(), "No Internet connection", Toast.LENGTH_LONG).show();
//                } catch (Exception e) {
//                    e.printStackTrace();
//                }
//            }
//            applyItemWeathers();
//        }
//    }
//    //
}

File: app/src/main/java/com/jeksonshar/jweather/ui/WeatherFragment.kt
package com.jeksonshar.jweather.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.jeksonshar.jweather.R
import com.jeksonshar.jweather.model.WeatherModel
import com.jeksonshar.jweather.repository.RepositoryDataBase
import com.jeksonshar.jweather.repository.RepositoryProvider
import java.text.SimpleDateFormat
import java.util.*

class WeatherFragment : Fragment() {

    private var city = 922137 // Kharkiv //TODO перенести во вьюмодель(бизнес логика)
                                        // планировал сделать запрос на получения номера вводимого города, выбирать город
    private var mRecyclerView: RecyclerView? = null
    private var mWeatherViewModel: WeatherViewModel? = null //TODO ViewModel должен быть NotNull
    private var lastUpdateView: TextView? = null
    private var changeView: ImageButton? = null
    private var mRepository: RepositoryDataBase? = null //TODO репозиторий во вьюмодель нужно перенести и обращаться к нему через viewModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mRepository = RepositoryProvider.getInstanceListWeather(context)
        mWeatherViewModel = ViewModelProvider(this, ViewModelFactory(city))
                .get(WeatherViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_weather, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        lastUpdateView = view.findViewById(R.id.last_update_view) // TODO переделать на синтетики или на ViewBinding обращения к полям
        mRecyclerView = view.findViewById(R.id.recyclerView)
        changeView = view.findViewById(R.id.fab)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {//TODO логику с полями вью фрагмента лучше перенести в onViewCreated
        super.onActivityCreated(savedInstanceState)

        var mWeatherAdapter: WeatherAdapter?
        var mWeatherModels: List<WeatherModel>?

        mRecyclerView!!.layoutManager = LinearLayoutManager(context)
        mWeatherModels = RepositoryProvider.getInstanceListWeather(context)?.allItemsWeather //TODO не видно зависимости в классе, обсудим на звонке

        if (mWeatherModels != null) {
            mWeatherAdapter = WeatherAdapter(mWeatherModels)
            mRecyclerView!!.adapter = mWeatherAdapter
        }

        var lastUpdate: String? = RepositoryProvider.getInstanceLastUpdate(context)?.lastUpdate
        lastUpdateView!!.text = lastUpdate
        mWeatherViewModel!!.request.observe(viewLifecycleOwner, Observer { weatherModels ->
            mWeatherModels = weatherModels

            if (mWeatherModels!!.isNotEmpty()) {//TODO в данном случае из viewModel нужно присылать notnull список
                val dateFormat = SimpleDateFormat(
                        " d MMM yyyy; HH:mm", Locale.getDefault())
                lastUpdate = dateFormat.format(WeatherViewModel.lastUpdate.time) //TODO непонятная логика с lastUpdate, надо убрать
                                         // в случае, если ответ на запрос получен, во вьюМодел получаем дату и время и форматируем
                lastUpdateView!!.text = lastUpdate
                RepositoryProvider.getInstanceLastUpdate(context)?.lastUpdate = lastUpdate
                mRepository!!.saveItemsInRoom(mWeatherModels)
            } else {
                Toast.makeText(context, R.string.no_internet, Toast.LENGTH_LONG)
                        .show() //TODO текст лучше вынести во вьюмодель
            }                   // не получается реализовать Toast в ViewModel, не могу получить там контекст

            mWeatherAdapter = RepositoryProvider //TODO обращение к репозиторию через viewModel
                    .getInstanceListWeather(context)
                    ?.allItemsWeather?.let { WeatherAdapter(it) }
            mRecyclerView!!.adapter = mWeatherAdapter
        })

        changeView!!.setOnClickListener {
            //TODO планирую реализовать смену отображения на график (х - число, у - значения параметров) и обратно
        }
    }
}